cs/src/core/Index/FasterLog/FasterLog.cs

```diff
@@ -0,0 +1,321 @@
+        {
+            epoch.Resume();
+            var length = entry.Length;
+            BlockAllocate(4 + length, out long logicalAddress);
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = entry)
+                Buffer.MemoryCopy(bp, (void*)(4 + physicalAddress), length, length);
+            epoch.Suspend();
+            return logicalAddress;
+        }
+
+        /// <summary>
+        /// Try to append entry to log
+        /// </summary>
+        /// <param name="entry">Entry to be appended to log</param>
+        /// <param name="logicalAddress">Logical address of added entry</param>
+        /// <returns>Whether the append succeeded</returns>
+        public unsafe bool TryAppend(byte[] entry, out long logicalAddress)
+        {
+            epoch.Resume();
+            logicalAddress = 0;
+            long tail = -allocator.GetTailAddress();
+            allocator.CheckForAllocateComplete(ref tail);
+            if (tail < 0)
+            { 
+                epoch.Suspend();
+                return false;
+            }
+            var length = entry.Length;
+            BlockAllocate(4 + length, out logicalAddress);
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = entry)
+                Buffer.MemoryCopy(bp, (void*)(4 + physicalAddress), length, length);
+            epoch.Suspend();
+            return true;
+        }
+
+        /// <summary>
+        /// Try to append entry to log
+        /// </summary>
+        /// <param name="entry">Entry to be appended to log</param>
+        /// <param name="logicalAddress">Logical address of added entry</param>
+        /// <returns>Whether the append succeeded</returns>
+        public unsafe bool TryAppend(Span<byte> entry, out long logicalAddress)
+        {
+            epoch.Resume();
+            logicalAddress = 0;
+            long tail = -allocator.GetTailAddress();
+            allocator.CheckForAllocateComplete(ref tail);
+            if (tail < 0)
+            {
+                epoch.Suspend();
+                return false;
+            }
+            var length = entry.Length;
+            BlockAllocate(4 + length, out logicalAddress);
+            var physicalAddress = allocator.GetPhysicalAddress(logicalAddress);
+            *(int*)physicalAddress = length;
+            fixed (byte* bp = &entry.GetPinnableReference())
+                Buffer.MemoryCopy(bp, (void*)(4 + physicalAddress), length, length);
+            epoch.Suspend();
+            return true;
+        }
+
+        /// <summary>
+        /// Append batch of entries to log
+        /// </summary>
+        /// <param name="entries"></param>
+        /// <returns>Logical address of last added entry</returns>
+        public unsafe long Append(List<byte[]> entries)
```

Comment:
+1 to having a batch append API. However this API forces too many allocations - 1 for the list, and 1 array allocation for each entry in this case:
If the serialized byte array is generated by a serializer that pools its' buffers (like Microsoft Bond), most likely it'll either return the result as an ArraySegment / Span / IMemoryOwner<byte>.

How does something like this look:

public unsafe long Append(int numEntries, Func<int, ArraySegment<byte>> getBytes)

(note that Span<byte> cannot be used here in place of ArraySegment<byte>, details here: https://adamsitnik.com/Span/#span-must-not-be-a-generic-type-argument)

Reply:
As an aside - Does the batch partially show up to the iterator while its still being written? Or does the reader only see the first event from the batch after epoch.Suspend() ?

Reply:
@hiteshmadan 
I don't understand everything thats going on in this PR, but if you want to use a span instead of an array segment you just need to declare a custom delegate like
```csharp
public delegate ReadOnlySpan<byte> ByteGetter<TState>(int index, TState state);
```
or something like it. That sidesteps the generic parameter constraint.

Reply:
After all the optimizations and inlining, there isn't much benefit to a batched interface in terms of performance, because acquire and release epoch - which were the only calls getting amortized - have been made very inexpensive. Also, we can't allocate for the batch in bulk because the page sizes are pretty small. Maybe we should stick to the unbatched interface?